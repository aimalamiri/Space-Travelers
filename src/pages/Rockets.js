import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Container } from 'react-bootstrap';
import Rocket from '../components/Rocket';

import { getRocketList } from '../redux/rockets/reducer';

const Rockets = () => {
  const rockets = useSelector((state) => state.rockets);
  const dispatch = useDispatch();
  useEffect(() => {
    if (rockets.length === 0) {
      dispatch(getRocketList());
    }
  }, []);

  return (
    <div className="rockets-main-container">
      <Container>
        {rockets.map((rocket) => (
          <Rocket rocket={rocket} key={rocket.id} />
        ))}
      </Container>
    </div>
  );
};

export default Rockets;
