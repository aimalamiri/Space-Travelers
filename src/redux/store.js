import { configureStore } from '@reduxjs/toolkit';
import thunk from 'redux-thunk';
import logger from 'redux-logger';
import missionsReducer from './missions/missions';
import rocketsReducer from './rockets/reducer';

export default configureStore({
  reducer: {
    missions: missionsReducer,
    rockets: rocketsReducer,
  },
  middleware: [logger, thunk],
});
