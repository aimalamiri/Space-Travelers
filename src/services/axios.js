import axiosInstance from 'axios';

export default axiosInstance.create({
  baseURL: 'https://api.spacexdata.com/v3/',
});
