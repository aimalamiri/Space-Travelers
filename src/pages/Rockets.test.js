import { Provider } from 'react-redux';
import {
  render, screen, waitFor, fireEvent,
} from '@testing-library/react';
import Rockets from './Rockets';
import MyProfile from './Profile';
import store from '../redux/store';
import axios from '../services/axios';

jest.mock('../services/axios');

describe('Test Rockets page', () => {
  beforeEach(async () => {
    const state = {
      data: [
        {
          id: 1,
          name: 'Falcon 1',
          description:
            'The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth.',
          flickr_images: ['https://imgur.com/DaCfMsj.jpg'],
        },
        {
          id: 2,
          name: 'Falcon 9',
          description:
            'Falcon 9 is a two-stage rocket designed and manufactured by SpaceX for the reliable and safe transport of satellites and the Dragon spacecraft into orbit.',
          flickr_images: ['https://farm1.staticflickr.com/929/28787338307_3453a11a77_b.jpg'],
        },
      ],
    };
    await axios.get.mockResolvedValue(state);
  });

  it('render the rockets page', async () => {
    render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );
    await waitFor(() => {
      expect(screen.getAllByText('Reserve Rocket').length).toBeGreaterThan(0);
    });
  });

  it('Marks a rocket as a reserved rocket when the user clicks the reserve button', async () => {
    render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );
    const reserveButtons = await screen.findAllByText('Reserve Rocket');
    fireEvent.click(reserveButtons[0]);
    const badges = await screen.findAllByText('Reserved');
    expect(badges.length).toBeGreaterThan(0);
  });

  it('Shows the reserved rocket in the profile page', async () => {
    render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );
    const reserveButtons = await screen.findAllByText('Reserve Rocket');
    fireEvent.click(reserveButtons[0]);
    render(
      <Provider store={store}>
        <MyProfile />
      </Provider>,
    );
    expect(screen.findByText('Falcon 1')).toBeTruthy();
  });

  it('Takes a snapshot', async () => {
    const tree = render(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    );
    await expect(tree).toMatchSnapshot();
  });
});
