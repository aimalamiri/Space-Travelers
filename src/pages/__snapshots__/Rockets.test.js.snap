// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Rockets page Takes a snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rockets-main-container"
      >
        <div
          class="container"
        >
          <div
            class="my-4 row"
            id="1"
          >
            <div
              class="col-md-3"
            >
              <img
                alt="rocket img"
                class="w-100"
                src="https://imgur.com/DaCfMsj.jpg"
              />
            </div>
            <div
              class="col"
            >
              <p
                class="rocket-name"
              />
              <p
                class="rocket-desc"
              >
                <span
                  class="reserve-span"
                >
                  Reserved
                </span>
                The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth.
              </p>
              <button
                class="btn btn-outline-secondary"
                type="button"
              >
                Cancel Reservation
              </button>
            </div>
          </div>
          <div
            class="my-4 row"
            id="2"
          >
            <div
              class="col-md-3"
            >
              <img
                alt="rocket img"
                class="w-100"
                src="https://farm1.staticflickr.com/929/28787338307_3453a11a77_b.jpg"
              />
            </div>
            <div
              class="col"
            >
              <p
                class="rocket-name"
              />
              <p
                class="rocket-desc"
              >
                <span
                  class="reserve-span"
                >
                  Reserved
                </span>
                Falcon 9 is a two-stage rocket designed and manufactured by SpaceX for the reliable and safe transport of satellites and the Dragon spacecraft into orbit.
              </p>
              <button
                class="btn btn-outline-secondary"
                type="button"
              >
                Cancel Reservation
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rockets-main-container"
    >
      <div
        class="container"
      >
        <div
          class="my-4 row"
          id="1"
        >
          <div
            class="col-md-3"
          >
            <img
              alt="rocket img"
              class="w-100"
              src="https://imgur.com/DaCfMsj.jpg"
            />
          </div>
          <div
            class="col"
          >
            <p
              class="rocket-name"
            />
            <p
              class="rocket-desc"
            >
              <span
                class="reserve-span"
              >
                Reserved
              </span>
              The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth.
            </p>
            <button
              class="btn btn-outline-secondary"
              type="button"
            >
              Cancel Reservation
            </button>
          </div>
        </div>
        <div
          class="my-4 row"
          id="2"
        >
          <div
            class="col-md-3"
          >
            <img
              alt="rocket img"
              class="w-100"
              src="https://farm1.staticflickr.com/929/28787338307_3453a11a77_b.jpg"
            />
          </div>
          <div
            class="col"
          >
            <p
              class="rocket-name"
            />
            <p
              class="rocket-desc"
            >
              <span
                class="reserve-span"
              >
                Reserved
              </span>
              Falcon 9 is a two-stage rocket designed and manufactured by SpaceX for the reliable and safe transport of satellites and the Dragon spacecraft into orbit.
            </p>
            <button
              class="btn btn-outline-secondary"
              type="button"
            >
              Cancel Reservation
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
